### ISSUE TITLE
Issues in mlpack MultiHeadAttention Layer Implementation

### ISSUE THREAD

**Issue by kumarutkarsh1248** [mlpack/mlpack#3969]
My current project on implementing the BERT architecture in mlpack and transferring weights from DistilBERT.onnx helped me explore the current implementation of mlpack's MultiHeadAttention layer. While going through the implementation, I noticed multiple issues that need to be resolved.

### 1. Attention Mask and Key Padding Mask Handling

The current implementation assumes a single attention mask and key padding mask, irrespective of batch size. This design is incorrect for batched processing, where each sentence in the batch should have its own mask.

#### Current implementation dimensions:
attentionMask: (tgtSeqLen, srcSeqLen)  
keyPaddingMask: (srcSeqLen)

#### Required dimensions for batched input:
attentionMask: cube (tgtSeqLen, srcSeqLen, batchSize)  
keyPaddingMask: matrix (srcSeqLen, batchSize)

### 2. Incorrect Softmax Axis in Key Padding Mask Application
Problematic Code Segment: [link](https://github.com/mlpack/mlpack/blob/82f4ff83ca369479c8e90d2fa499344de9e1fbec/src/mlpack/methods/ann/layer/multihead_attention_impl.hpp#L159-L162)  
scores is a cube with dimensions (qProj, kProj, batchSize). After adding the key padding mask and applying softmax, we expect certain keys to be masked—meaning some rows (corresponding to specific keys) should become zeroed out. However, the problem is that mlpack::Softmax::Forward() applies softmax along the column elements, whereas here, we need to apply softmax along the row elements for correct masking behavior.
To resolve this just some transpose is required and after changing the code with thi
```cpp
  for (size_t i = 0; i < numHeads * batchSize; ++i)
  {
    softmax.Forward(scores.slice(i).t(), scores.slice(i));
    scores.slice(i) = scores.slice(i).t();
  }
```

i am getting desired behaviour

### 3. Incorrect Dimension Handling During Attention Output Projection

Problematic Code Segment: [link](https://github.com/mlpack/mlpack/blob/82f4ff83ca369479c8e90d2fa499344de9e1fbec/src/mlpack/methods/ann/layer/multihead_attention_impl.hpp#L173-L178)  
In the final projection step of the attention output, matrix multiplication is performed along incorrect dimensions
#### Solution:
Use transpose before multiplication:

```cpp
for (size_t i = 0; i < batchSize; ++i)
{
  output.col(i) = vectorise( outWt * trans(attnOut.slice(i)) +
      trans(repmat(outBias, tgtSeqLen, 1))
  );
}
```

**Comment by rcurtin** [comment]
Thanks for the report @kumarutkarsh1248 and sorry it took me a little bit to get back to it.  CC @akropp in case you have thoughts on the issues here too. 

1. I agree with you here, do you want to open a PR for the issue?
 
2. The fix you proposed does work, but the double transpose will add significant slowdown to the forward pass.  It would be better if we could avoid that... and it seems like that could be possible if we just changed `scores`:
```
scores = MultiplyCube2Cube(qProj, kProj, false, true);
```

If you switch `kProj` and `qProj` here (and update the transpose flags) then we get `scores` as shape `(kProj, qProj, batchSize)` and the softmax layer will apply correctly as-is and avoid the transpose.

As a side note on this point, note that [Armadillo is column-major](https://www.mlpack.org/doc/user/matrices.html) but typical machine learning math is row-major; so the multihead attention softmax operation is (in the paper) `softmax( QK^T / sqrt(d_k) ) V`, but that assumes that each query is a row in `Q`.  In our case each query is a column (same with `K` and `V`), so instead what we need to implement for that equation is really `V softmax_column_major( K Q^T / sqrt(d_k) )^T`.

3. I agree, I think this is related to the point above about column-major operations.  So with the changes above, `attnOut` will end up in a transposed form.  You can also avoid the `trans()` on the `repmat()` just by switching the second and third arguments (might need to also reshape `outBias` in other parts of the code).

**Comment by akropp** [comment]
Not having looked at this for a while, I think what @rcurtin suggested sounds right as far as creating scores transposed.  You have to take care with the masks as well (the attnMask and keyPaddingMask will have to be added to scores transposed).  

Then when multiplying by vProj, you will have to transpose scores (just pass aTranspose=true and reverse scores and  @vProj), which will result in attnOut being the same dims it is currently (`tgtSeqLen x numHeads x (numHeads*batchSize`) ->  (`tgtSeqLen x embedDim x batchSize`).  

Not sure about the transpose in the final output.  shouldn't the current implementation be the same as what is proposed?  attnOut.slice(i) is dimensions `(tgtSeqLen x embedDim)` and outWt is `(embedDim x embedDim)`,so `trans(attnOut.slice(i) * outWt)` is dimension `(embedDim x tgtSeqLen)`, same as `outWt * trans(attnOut.slice(i))`.  So all depends on how the weights in outWt are represented/used elsewhere (column-wise or row-wise).

**Comment by rcurtin** [comment]
Fixed in #3998.
